gnd/src/main/java/com/google/android/gnd/workers/TileDownloadWorker.java

```diff
@@ -42,106 +45,106 @@
-        }
-      }
-
-      connection.connect();
-
-      InputStream is = connection.getInputStream();
-      FileOutputStream fos = context.openFileOutput(tile.getPath(), Context.MODE_PRIVATE);
-      byte[] byteChunk = new byte[BUFFER_SIZE];
-      int n;
-
-      while ((n = is.read(byteChunk)) > 0) {
-        fos.write(byteChunk, 0, n);
+  private void downloadTileFile(Tile tile, Map<String, String> requestProperties)
+      throws IOException {
+    URL url = new URL(tile.getUrl());
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+
+    if (!requestProperties.isEmpty()) {
+      for (Map.Entry<String, String> property : requestProperties.entrySet()) {
+        connection.setRequestProperty(property.getKey(), property.getValue());
       }
+    }
 
-      is.close();
-      fos.close();
-
-      localDataStore
-          .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.DOWNLOADED).build())
-          .blockingAwait();
-
-      return Result.success();
-
-    } catch (IOException e) {
-      Log.d(TAG, "Failed to download and write file.", e);
+    connection.connect();
 
-      localDataStore
-          .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.FAILED).build())
-          .blockingAwait();
+    InputStream is = connection.getInputStream();
+    FileOutputStream fos = context.openFileOutput(tile.getPath(), Context.MODE_PRIVATE);
+    byte[] byteChunk = new byte[BUFFER_SIZE];
+    int n;
 
-      return Result.failure();
+    while ((n = is.read(byteChunk)) > 0) {
+      fos.write(byteChunk, 0, n);
     }
-  }
-
-  /** Update a tile's state in the database and initiate a download of the tile source file. */
-  private Result downloadTile(Tile tile) {
-    localDataStore
-        .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.IN_PROGRESS).build())
-        .blockingAwait();
 
-    return downloadTileFile(tile, Optional.empty());
+    is.close();
+    fos.close();
   }
 
-  /** Resumes downloading the source for {@param tile} marked as {@code Tile.State.IN_PROGRESS}. */
-  private Result resumeTileDownload(Tile tile) {
-    File existingTileFile = new File(context.getFilesDir(), tile.getPath());
+  /** Update a tile's state in the database and initiate a download of the tile source file. */
+  private Completable downloadTile(Tile tile) {
     HashMap<String, String> requestProperties = new HashMap<>();
 
-    requestProperties.put("Range", existingTileFile.length() + "-");
+    if (tile.getState() == State.IN_PROGRESS) {
+      File existingTileFile = new File(context.getFilesDir(), tile.getPath());
+      requestProperties.put("Range", existingTileFile.length() + "-");
+    }
 
-    return downloadTileFile(tile, Optional.of(requestProperties));
+    return localDataStore
+        .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.IN_PROGRESS).build())
+        .andThen(
+            Completable.fromCallable(
```

Comment:
If you use `Completable.fromRunnable` instead you don't need to return anything and can use an inline lambda instead of a block.

Reply:
I wanted to do this, but since the IO exception here is a checked exception, we'd have to explicitly handle the error in the runnable itself, losing the elegance of its automatic propagation upstream (though I suppose NBD since we're already losing elegance with this approach). see here for more info: https://stackoverflow.com/questions/44179444/how-to-create-observable-out-of-runnable

Our options are what we currently have here, or:

```java
Completable.fromRunnable(() -> {try {downloadTile(tile)} catch (IOException e) { Log.d(TAG, "Failed to download tile: " + tile, e);}})
``` 

I opted for the Callable variant since it's shorter. Though I think it may be more confusing to someone unfamiliar with the checked exception restriction....but then again so would handling the exception in the runnable itself.

I'm good with using whichever style you prefer.

Reply:
Perhaps `downloadTileFile()` should catch the error and rethrow an unchecked exception?

We need to try {} finally {} in `downloadTileFile()` to close the connection, so adding `catch()` should be easy. Wdyt?

Reply:
> Perhaps `downloadTileFile()` should catch the error and rethrow an unchecked exception?
> 
> We need to try {} finally {} in `downloadTileFile()` to close the connection, so adding `catch()` should be easy. Wdyt?

That's a great idea! Let's do that.

Reply:
Done see 2b81c0b